src/coreclr/nativeaot/Runtime/unix/PalCreateDump.cpp

```diff
@@ -0,0 +1,583 @@
+#include <sys/param.h>
+#include <sys/sysctl.h>
+#include <kvm.h>
+#endif
+
+#ifdef __FreeBSD__
+#include <sys/sysctl.h>
+#include <sys/user.h>
+#endif
+
+#define _T(s) s
+#include "RhConfig.h"
+
+#include <minipal/utils.h>
+#include <generatedumpflags.h>
+
+// Crash dump generating program arguments. MAX_ARGV_ENTRIES is the max number
+// of entries if every createdump option/argument is passed.
+#define MAX_ARGV_ENTRIES 32 
+const char* g_argvCreateDump[MAX_ARGV_ENTRIES] = { nullptr };
+char* g_szCreateDumpPath = nullptr;
+char* g_ppidarg  = nullptr;
+
+/*++
+Function:
+    PlatformGetCurrentThreadId
+
+    Returns the current thread id
+*/
+
+#if defined(__linux__)
+#define PlatformGetCurrentThreadId() (uint32_t)syscall(SYS_gettid)
+#elif defined(__APPLE__)
+inline uint32_t PlatformGetCurrentThreadId() {
+    uint64_t tid;
+    pthread_threadid_np(pthread_self(), &tid);
+    return (uint32_t)tid;
+}
+#elif defined(__FreeBSD__)
+#include <pthread_np.h>
+#define PlatformGetCurrentThreadId() (uint32_t)pthread_getthreadid_np()
+#elif defined(__NetBSD__)
+#include <lwp.h>
+#define PlatformGetCurrentThreadId() (uint32_t)_lwp_self()
+#else
+#define PlatformGetCurrentThreadId() (uint32_t)pthread_self()
+#endif
+
+const size_t MaxUnsigned32BitDecString = ARRAY_SIZE("4294967295") - 1;
+const size_t MaxUnsigned64BitDecString = ARRAY_SIZE("18446744073709551615") - 1;
+
+/*++
+Function:
+    FormatInt
+
+    Helper function to format an uint32 as a string.
+
+--*/
+static
+char*
+FormatInt(uint32_t value)
+{
+    char* buffer = (char*)malloc(MaxUnsigned32BitDecString);
+    if (buffer != nullptr)
+    {
+        if (snprintf(buffer, MaxUnsigned32BitDecString, "%" PRIu32, value) < 0)
+        {
+            free(buffer);
+            buffer = nullptr;
+        }
+    }
+    return buffer;
+}
+
+/*++
+Function:
+    FormatInt64
+
+    Helper function to format an uint64 as a string.
+
+--*/
+static
+char*
+FormatInt64(uint64_t value)
+{
+    char* buffer = (char*)malloc(MaxUnsigned64BitDecString);
+    if (buffer != nullptr)
+    {
+        if (snprintf(buffer, MaxUnsigned64BitDecString, "%" PRIu64, value) < 0)
+        {
+            free(buffer);
+            buffer = nullptr;
+        }
+    }
+    return buffer;
+}
+
+static const int UndefinedDumpType = 0;
+
+/*++
+Function
+  BuildCreateDumpCommandLine
+
+Abstract
+  Builds the createdump command line from the arguments.
+
+Return
+  true - succeeds, false - fails
+
+--*/
+static
+bool
+BuildCreateDumpCommandLine(
+    const char** argv,
+    const char* dumpName,
+    const char* logFileName,
+    int dumpType,
+    uint32_t flags)
+{
+    if (g_szCreateDumpPath == nullptr || g_ppidarg == nullptr)
+    {
+        return false;
+    }
+
+    int argc = 0;
+    argv[argc++] = g_szCreateDumpPath;
+
+    if (dumpName != nullptr)
+    {
+        argv[argc++] = "--name";
+        argv[argc++] = dumpName;
+    }
+
+    switch (dumpType)
+    {
+        case 1: argv[argc++] = "--normal";
```

Comment:
Use the constants here - we could use [generatedumpflags.h](https://github.com/dotnet/runtime/pull/89203/files#diff-fcf0912fe21b30ae2fea3d5be8b0118d78494ba7029e7c07704038d5ef1557c5) as a shared header for constants between runtime and createdump